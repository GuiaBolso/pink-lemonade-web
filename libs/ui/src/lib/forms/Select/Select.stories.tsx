import React from 'react';
import { Select as S } from './Select';

export default {
  title: 'Forms/Select',
  component: S,
};

const options = [
  {
    label: 'lorem 1',
    value: '1',
  },
  {
    label: 'lorem 2',
    value: '2',
  },
];

export const Select = () => <S options={options} label="Isto é uma label" />;

export const error = () => (
  <S options={options} label="Isto é uma label" error errorMessage="Error" />
);
