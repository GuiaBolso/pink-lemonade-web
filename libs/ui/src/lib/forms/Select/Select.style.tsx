import styled from '@emotion/styled';
import { pxToRem } from '@guiabolsobr/utils';
import { CustomThemeProps } from '../../../typings/CustomThemeProps';

type Theme = {
  theme: CustomThemeProps;
  error?: boolean;
};

export const WrapperSelect = styled.div`
  & > .MuiFormControl-root {
    font-family: inherit;
    height: ${pxToRem(64)};
    width: 100%;

    .MuiPopover-paper {
      top: 84px !important;
    }

    &.Mui-focused,
    .MuiFormLabel-root.Mui-focused {
      color: ${({ theme }: Theme) => theme?.colors?.neutral?.medium} !important;
    }

    &:focus:before {
      border-color: ${({ theme, error }: Theme) =>
        error
          ? theme?.colors?.feedback?.error?.default
          : theme?.colors?.brand?.tertiary?.default};
    }

    &:hover {
      background: none;

      &:before {
        border-color: ${({ theme, error }: Theme) =>
          error
            ? theme?.colors?.feedback?.error?.default
            : theme?.colors?.brand?.tertiary?.default};
        border-width: ${pxToRem(2)};
      }

      & .MuiSelect-icon {
        color: ${({ theme }: Theme) => theme?.colors?.brand?.tertiary?.default};
        top: calc(50% - 14px);
      }
    }

    &:before {
      animation: border 0.3s linear;
      border: 1px solid
        ${({ theme, error }: Theme) =>
          error
            ? theme?.colors?.feedback?.error?.default
            : theme?.colors?.neutral?.regular};
      border-radius: 2px;
      bottom: 0;
      content: '';
      left: 0;
      padding: 1px;
      pointer-events: none;
      position: absolute;
      right: 0;
      top: 0;
    }

    & .MuiSelect-iconFilled {
      right: 23px;
    }

    & .MuiSvgIcon-root {
      font-size: 2rem;
    }

    & .MuiSelect-icon {
      color: ${({ theme }: Theme) => theme?.colors?.brand?.secondary?.default};
      top: calc(50% - 14px);
    }

    & .MuiSelect-select:focus {
      background: none;
    }

    & .MuiFormLabel-root.Mui-error {
      color: ${({ theme }: Theme) => theme?.colors?.neutral?.dark};
    }

    & .MuiSelect-filled.MuiSelect-filled {
      color: ${({ theme }: Theme) => theme?.colors?.neutral?.dark};
      font-size: ${pxToRem(18)};
      line-height: ${pxToRem(22)};
      padding-left: ${pxToRem(16)};
    }

    & .MuiFilledInput-root {
      background: none;
      border-top-left-radius: 0px;
      border-top-right-radius: 0px;
      position: relative;
    }

    & .MuiFilledInput-underline {
      &:before {
        border-bottom: 0;
      }

      &:hover:before {
        border-bottom: 0;
      }
    }

    & .MuiInputLabel-formControl {
      font-size: ${pxToRem(18)};
      left: 5px;
      line-height: ${pxToRem(22)};
      transform: translate(12px, 22px) scale(1);
    }

    & .Mui-focused:after,
    & .MuiFilledInput-underline.Mui-error:after {
      transform: scaleX(0);
    }

    & .MuiFormHelperText-root.Mui-error {
      margin-left: 0;
      margin-top: 9px;
    }

    & .MuiFormHelperText-root.Mui-error,
    & .MuiSelect-icon {
      color: ${({ theme }: Theme) => theme?.colors?.feedback?.error?.default};
    }

    .MuiInputLabel-filled.MuiInputLabel-shrink {
      transform: translate(12px, 10px) scale(0.7) !important;
    }
  }
`;
