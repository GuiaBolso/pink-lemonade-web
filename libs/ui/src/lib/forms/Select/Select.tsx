import React from 'react';
import { WrapperSelect } from './Select.style';
import FormControl from '@material-ui/core/FormControl';
import InputLabel from '@material-ui/core/InputLabel';
import FormHelperText from '@material-ui/core/FormHelperText';
import {
  default as SelectMui,
  SelectProps as SProps,
} from '@material-ui/core/Select';

import * as S from '../../navigation/DropMenu/DropMenu.style';

export type SelectProps = {
  options: Array<{ label: string | number | React.ReactNode; value: string }>;
  error?: boolean;
  errorMessage?: string;
} & SProps;

export const Select = ({
  options,
  label,
  error,
  errorMessage,
  ...props
}: SelectProps) => {
  return (
    <WrapperSelect error={error}>
      <FormControl variant="filled" error={error}>
        <InputLabel>{label}</InputLabel>
        <SelectMui
          variant="filled"
          label={label}
          MenuProps={{
            disablePortal: true,
          }}
          {...props}
        >
          {options.map(option => (
            <S.MenuItem value={option.value} key={`select-${option.value}`}>
              {option.label}
            </S.MenuItem>
          ))}
        </SelectMui>
        {errorMessage && <FormHelperText>{errorMessage}</FormHelperText>}
      </FormControl>
    </WrapperSelect>
  );
};
