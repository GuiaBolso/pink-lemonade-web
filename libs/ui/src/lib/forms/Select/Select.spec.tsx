import React from 'react';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { Select } from './Select.stories';

describe('Select Component', () => {
  it('Snapshot', () => {
    expect(renderer.create(<Select />).toJSON()).toMatchSnapshot();
  });
});
