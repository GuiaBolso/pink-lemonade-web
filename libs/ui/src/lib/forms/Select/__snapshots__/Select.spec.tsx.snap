// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component Snapshot 1`] = `
<div
  className="css-1orwgc1"
>
  <div
    className="MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
      data-shrink={false}
    >
      Isto Ã© uma label
    </label>
    <div
      className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-formControl"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-filled MuiInputBase-input MuiFilledInput-input"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "&#8203;",
            }
          }
        />
      </div>
      <input
        aria-hidden={true}
        className="MuiSelect-nativeInput"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconFilled"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</div>
`;
